### PR TITLE
[Scheduler] Add new component documentation

### LINKED ISSUE #18067: [Messenger] Add a scheduler component
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#47112
| PR author(s) | @upyx, @fabpot
| Merged in    | 6.3

We created this issue to not forget to document this new feature. We would really appreciate if you can help us with this task. If you are not sure how to do it, please ask us and we will help you.

To fix this issue, please create a PR against the 6.3 branch in the [symfony-docs repository](https://github.com/symfony/symfony-docs).

Thank you! :smile:

### PR DESCRIPTION
**Pull request by andreybolonin**
Fixes #18067, #18255, #18251

Based on #18136

Waiting for feedbacks and ready to fix some issues :)

### CODE REVIEW

File: scheduler.rst [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+--------------------------
+
+Scheduler uses the same :ref:`Messenger worker <messenger-worker>` to consume
+messages dispatched by your schedule class.
+
+You can do this with the ``messenger:consume`` command:
+
+.. code-block:: terminal
+
+    $ php bin/console messenger:consume scheduler_trial
```
Comment (alexisLefebvre) [inline scheduler.rst:62]: Where do `scheduler_trial` comes from?

File: scheduler.rst [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+---------------------
+
+Symfony provides by default two triggers for your schedules : `every` and
+`cron`. By extending the provided TriggerInterface you can also create
+your own trigger for your app.
+
+Every
+~~~~~
+
+This trigger allows you to define you schedule with frequency as text, using
```
Comment (Aerendir) [inline scheduler.rst]: "allows you to define YOUR"

File: scheduler.rst [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    RecurringMessage::every('first day of month', $msg, from: '13:47');
+
+    // using timezone
+    RecurringMessage::every('first day of month', $msg, from: '13:47+0400');
+
+    // using a DateTimeImmutable object
+    $from = new \DateTimeImmutable('13:47', new \DateTimeZone('Europe/Paris'));
+    RecurringMessage::every('first day of month', $msg, from: $from);
+
+If your schedule do not need to run indefinitely, you can use `until` option
```
Comment (Aerendir) [inline scheduler.rst]: - "If your schedule DOES not"
- "you can use THE `until`"

File: scheduler.rst [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+to let the Component know when the trigger should stop::
+
+    RecurringMessage::every('monday', $msg, until: '2023-06-12');
+
+Cron
+~~~~
+
+This trigger can be configured following same rules as the eponymous Unix
```
Comment (Aerendir) [inline scheduler.rst]: "following THE same rules OF the"

File: scheduler.rst [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+    $context->nextTriggerAt; //22:14:00
+    
+    $iterator->next();
+    
+    $context->triggeredAt; //22:14:00
+
+    $context->nextTriggerAt; //22:16:00
+    
+Also you can use some another triggers: ``CallbackTrigger, CronExpressionTrigger, DateIntervalTrigger, DatePeriodTrigger, ExcludeTimeTrigger``
```
Comment (Aerendir) [inline scheduler.rst]: "You can also use some other triggers"

File: scheduler.rst [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+Define your own trigger
+~~~~~~~~~~~~~~~~~~~~~~~
+
+By implementing the TriggerInterface your can create your own trigger using
```
Comment (Aerendir) [inline scheduler.rst]: "By implementing the TriggerInterface, you can create your own trigger using

**Review by Aerendir — CHANGES_REQUESTED** [review]

File: scheduler.rst [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+intervals or relative date format allowed by PHP::
+
+    RecurringMessage::every('10 seconds', $msg);
+    RecurringMessage::every('1 day', $msg);
+    RecurringMessage::every('2 weeks', $msg);
+
+    RecurringMessage::every('first day of month', $msg);
+    RecurringMessage::every('next tuesday', $msg);
+
+It is possible to specify an exact hour to trigger the schedule with `from`
```
Comment (alexandre-daubois) [inline scheduler.rst:84]:
```suggestion
It is possible to specify an exact hour to trigger the schedule with ``from``
```

File: scheduler.rst [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    RecurringMessage::every('first day of month', $msg, from: '13:47');
+
+    // using timezone
+    RecurringMessage::every('first day of month', $msg, from: '13:47+0400');
+
+    // using a DateTimeImmutable object
+    $from = new \DateTimeImmutable('13:47', new \DateTimeZone('Europe/Paris'));
+    RecurringMessage::every('first day of month', $msg, from: $from);
+
+If your schedule does not need to run indefinitely, you can use the `until` option
```
Comment (alexandre-daubois) [inline scheduler.rst:96]:
```suggestion
If your schedule does not need to run indefinitely, you can use the ``until`` option
```

File: scheduler.rst [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+    // trigger every hour
+    RecurringMessage::cron('0 * * * *', $msg);
+
+    // trigger every monday at 12:00
+    RecurringMessage::cron('0 12 * * 1', $msg);
+
+.. note::
+
+    The minimal interval allowed with cron is 1 minute.
```
Comment (alexandre-daubois) [inline scheduler.rst:115]:
```suggestion
    The minimal interval allowed by the ``cron()`` method is 1 minute.
```

File: scheduler.rst [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    {
+        public function getSchedule(string $id): Schedule
+        {
+            return (new Schedule())
+                ->add(new RecurringMessage::every('1 week', new EndofTrialMessage()))
+            ;
+        }
+    }
+
+Run the Messenger consumer
```
Comment (alexandre-daubois) [inline scheduler.rst:52]:
```suggestion
Run The Messenger Consumer
```

File: scheduler.rst [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+Scheduler uses the same :ref:`Messenger worker <messenger-worker>` to consume
+messages dispatched by your schedule class.
+
+You can do this with the ``messenger:consume`` command:
+
+.. code-block:: terminal
+
+    $ php bin/console messenger:consume scheduler_trial
+
+Trigger your schedule
```
Comment (alexandre-daubois) [inline scheduler.rst:64]:
```suggestion
Trigger Your Schedule
```

File: scheduler.rst [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+
+    // trigger every monday at 12:00
+    RecurringMessage::cron('0 12 * * 1', $msg);
+
+.. note::
+
+    The minimal interval allowed with cron is 1 minute.
+
+Hashed expressions
+~~~~
```
Comment (alexandre-daubois) [inline scheduler.rst:118]:
```suggestion
~~~~~~~~~~~~~~~~~~
```

File: scheduler.rst [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    RecurringMessage::cron('#daily', $msg);
+
+A hashed expressions is a string representing the schedule for a particular command to execute.
+Supported hashed expression:
+
+    * `#yearly`, `#annually` - Run once a year, midnight, Jan. 1 - `0 0 1 1 *`
+    * `#monthly` - Run once a month, midnight, first of month - `0 0 1 * *`
+    * `#weekly` - Run once a week, midnight on Sun - `0 0 * * 0`
+    * `#daily`, `#midnight` - Run once a day, midnight - `0 0 * * *`
+    * `#hourly` - Run once an hour, first minute - `0 * * * *`
```
Comment (alexandre-daubois) [inline scheduler.rst:138]:
```suggestion
    * ``#yearly``, ``#annually`` - Run once a year, midnight, Jan. 1 - ``0 0 1 1 *``
    * ``#monthly`` - Run once a month, midnight, first of month - ``0 0 1 * *``
    * ``#weekly`` - Run once a week, midnight on Sun - ``0 0 * * 0``
    * ``#daily``, ``#midnight`` - Run once a day, midnight - ``0 0 * * *``
    * ``#hourly`` - Run once an hour, first minute - ``0 * * * *``
```

File: scheduler.rst [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+Message generator
+~~~~
```
Comment (alexandre-daubois) [inline scheduler.rst:141]:
```suggestion
~~~~~~~~~~~~~~~~~
```

File: scheduler.rst [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+
+    $iterator->next();
+
+    $context->triggeredAt; //22:14:00
+
+    $context->nextTriggerAt; //22:16:00
+
+You can also use some another triggers: ``CallbackTrigger, CronExpressionTrigger, DateIntervalTrigger, DatePeriodTrigger, ExcludeTimeTrigger``
+
+Define your own trigger
```
Comment (alexandre-daubois) [inline scheduler.rst:169]:
```suggestion
Define Your Own Trigger
```

File: scheduler.rst [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~
+
+By implementing the TriggerInterface, you can create your own trigger using
+``getNextRunDate`` method::
```
Comment (alexandre-daubois) [inline scheduler.rst:173]:
```suggestion
the :method:`FQCN\TriggerInterface::getNextRunDate` method::
```

**Review by alexandre-daubois — CHANGES_REQUESTED** [review]
Great to see this component being documented! :smiley:

**Comment by OskarStark** [comment]
@kbond maybe you can find some spare time and review this PR, so we can merge this as a first version. Thanks

**Comment by javiereguiluz** [comment]
**Please, let's not work on this PR anymore for now.**

It's not bad 🙏 but we're discussing internally about a radically different approach for the docs of this component. Same comment for the #17964 PR about Webhook/RemoteEvent docs.

I'm sorry but I hope you understand this. Thanks!

**Comment by Aerendir** [comment]
@javiereguiluz , please, can you give us more details? I'm so curious! 😋

**Pull request closed without merging**